core/src/main/java/com/linecorp/armeria/internal/client/ClientUtil.java

```diff
@@ -39,41 +42,77 @@ O initContextAndExecuteWithFallback(
             U delegate,
             DefaultClientRequestContext ctx,
             EndpointGroup endpointGroup,
-            BiFunction<ClientRequestContext, Throwable, O> fallbackResponseFactory) {
+            Function<CompletableFuture<O>, O> futureConverter,
+            BiFunction<ClientRequestContext, Throwable, O> errorResponseFactory) {
 
         requireNonNull(delegate, "delegate");
         requireNonNull(ctx, "ctx");
         requireNonNull(endpointGroup, "endpointGroup");
-        requireNonNull(fallbackResponseFactory, "fallbackResponseFactory");
+        requireNonNull(futureConverter, "futureConverter");
+        requireNonNull(errorResponseFactory, "errorResponseFactory");
 
         try {
             endpointGroup = mapEndpoint(ctx, endpointGroup);
-            if (ctx.init(endpointGroup)) {
-                return pushAndExecute(delegate, ctx);
-            } else {
-                // Context initialization has failed, which means:
-                // - ctx.log() has been completed with an exception.
-                // - ctx.request() has been aborted (if not null).
-                // - the decorator chain was not invoked at all.
-                // See `init()` and `failEarly()` in `DefaultClientRequestContext`.
-
-                // Call the decorator chain anyway so that the request is seen by the decorators.
-                final O res = pushAndExecute(delegate, ctx);
-
-                // We will use the fallback response which is created from the exception
-                // raised in ctx.init(), so the response returned can be aborted.
-                if (res instanceof StreamMessage) {
-                    ((StreamMessage<?>) res).abort();
+            final CompletableFuture<Boolean> initFuture = ctx.init(endpointGroup);
+            if (initFuture.isDone()) {
+                // Initialization has been done immediately.
+                final boolean success;
+                try {
+                    success = initFuture.get();
+                } catch (Exception e) {
+                    throw Exceptions.peel(e);
```

Comment:
Wrap with `UnprocessedRequestException.wrap(e)` once #2836 is merged?

Reply:
Done.